tests/python/integration_test.py

```diff
@@ -0,0 +1,206 @@
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+"""
+Provides a base class for integration tests.
+"""
+
+from __future__ import print_function
+
+import os
+import tempfile
+import atexit
+from mock import Mock
+
+import unittest2
+
+import sgtk
+from sgtk.util import sgre as re
+from sgtk.util.filesystem import safe_delete_folder
+import tk_toolchain.authentication
+
+
+class DesktopServerIntegrationTest(unittest2.TestCase):
+    """
+    Base class for integration tests. Each integration test should be invoke in its own subprocess.
+
+    The base class takes care of:
+        - setting up a log file named after the test
+        - creating a random temporary folder to write to or use the path pointed by SHOTGUN_TEST_TEMP
+        - setting SHOTGUN_HOME to point to <temp_dir>/shotgun_home
+        - authenticating a user using tk-toolchain
+        - cleaning up the test folder when the tests are done running.
+    """
+
+    @classmethod
+    def setUpClass(cls):
+        """
+        Sets up the test suite.
+        """
+
+        # Set up logging
+        sgtk.LogManager().initialize_base_file_handler(
+            cls._camel_to_snake(cls.__name__)
+        )
+        sgtk.LogManager().initialize_custom_handler()
+
+        # Create a temporary directory for these tests and make sure
+        # it is cleaned up.
+        if "SHOTGUN_TEST_TEMP" not in os.environ:
+            cls.temp_dir = tempfile.mkdtemp()
+            # Only clean up the temp dir when not retrieving coverage for the tests,
+            # or we won't be able ot merge the reports from all runs.
+            if "SHOTGUN_TEST_COVERAGE" not in os.environ:
+                # Do not rely on tearDown to cleanup files on disk. Use the atexit callback which is
+                # much more reliable.
+                atexit.register(cls._cleanup_temp_dir)
+        else:
+            cls.temp_dir = os.environ["SHOTGUN_TEST_TEMP"]
+
+        # Ensures calls to the tempfile module generate paths under the unit test temp folder.
+        tempfile.tempdir = cls.temp_dir
+
+        # Ensure Toolkit writes to the temporary directory
+        os.environ["SHOTGUN_HOME"] = os.path.join(cls.temp_dir, "shotgun_home")
+
+        # Create a user and connection to Shotgun.
+        sa = sgtk.authentication.ShotgunAuthenticator()
+        cls.user = tk_toolchain.authentication._get_toolkit_user(sa, os.environ)
+        cls.sg = cls.user.create_sg_connection()
+
+        # Set the current user in memory for the server and in the file system for the script that
+        # runs the actions in a separate process
+        sgtk.set_authenticated_user(cls.user)
+        sgtk.authentication.session_cache.set_current_host(cls.user.host)
```

Comment:
Ah! I think you're calling this (private) API so that the subprocess knows which user it is right? That's because there's a design flaw in how credentials are passed down to the subprocess. Could you fix that?

The [arguments file](https://github.com/shotgunsoftware/tk-framework-desktopserver/blob/master/python/tk_framework_desktopserver/shotgun/api_v2.py#L230) should pass along the credentials for the user and the `cache_commands` and `execute_command` should read the user from that file.

This way you do not have to rely on a global "variable". [Globals.... BAAAAAADDDD!](https://www.youtube.com/watch?time_continue=125&v=fS6udST6lbE&feature=emb_logo)

Reply:
You can use use [`serialize_user` and `deserialize_user`](https://developer.shotgunsoftware.com/tk-core/authentication.html#sgtk.authentication.deserialize_user).